client/checkout/blocks/upe-fields.js

```diff
@@ -381,10 +383,9 @@ const ConsumableWCPayFields = ( { api, ...props } ) => {
 		if ( paymentIntentId || hasRequestedIntent ) {
 			return;
 		}
-
-		async function createIntent() {
+		async function createIntent( paymentMethodId ) {
 			try {
-				const response = await api.createIntent();
+				const response = await api.createIntent( paymentMethodId );
```

Comment:
In the classic shortcode checkout, before creating the intent, we first check to see if [there is a cached payment intent on the current session](https://github.com/Automattic/woocommerce-payments/blob/fix/4603-block-split-upe/client/checkout/classic/upe.js#L231-L233). There should be a unique cache key for each payment method type at checkout, to mitigate the multiplications of payment intent requests that we make, ensuring that we will create at most a single payment intent at checkout for each UPE payment gateway.

In this blocks implementation, we don't appear to be doing this, which is exacerbated by the fact that every time we select a new radio button gateway at checkout, the _entire_ gateway re-renders, recreating and remounting a new payment element, _and_ creating a brand new payment intent as well! I found that if I simply switched back and forth between UPE gateways at checkout, watching my network tab, I could see that we created a brand new payment intent each time the element re-rendered.

I believe you should be able to pull the cached intent from the same UPE `paymentMethodsConfig` object that you're using here ([here's](https://github.com/Automattic/woocommerce-payments/blob/fix/4603-block-split-upe/includes/payment-methods/class-upe-payment-gateway.php#L1079-L1080) where that cached data gets stored in the config). Let me know if you think that'd be an easy update or if you tried it already and something unfortunate and troubling happened instead! 😬

Reply:
I did some minor research to find answers to our questions, which is why this comment came out a bit later. There is a slight difference in Stripe elements management between the classic and blocks checkout.

### Classic checkout
Stripe elements mounting [on the ](https://github.com/Automattic/woocommerce-payments/blob/88be4a965c11235ab78709577eba83967a48f09d/client/checkout/classic/upe.js#L352-L376)classic checkout page occur only once on the updated_checkout[ event](https://github.com/Automattic/woocommerce-payments/blob/88be4a965c11235ab78709577eba83967a48f09d/client/checkout/classic/upe.js#L352-L376). At this stage, the state is examined to confirm whether the UPE element was already mounted. If it was mounted, we ensure that the UI focuses on the element. If it wasn’t, we would utilize the mountUPEElement function. 

The mountUPEElement function, among others, [checks if there is no payment intent data related to the particular method in the session storage](https://github.com/Automattic/woocommerce-payments/blob/fix/4603-block-split-upe/client/checkout/classic/upe.js#L231-L233), as you have pointed out before. If there is nothing in the session storage, we contribute to creating a new payment intent with the help of the createIntent Stripe API method. 

The key point here to pay attention to is that [the attempt to fetch the payment intent data from the session](https://github.com/Automattic/woocommerce-payments/blob/fix/4603-block-split-upe/client/checkout/classic/upe.js#L231-L233) is not executed when users are switching back and forth between UPE gateways at the checkout page. Instead, it is executed when the page is refreshed. 

In other words, assuming that our session storage is empty (site cookies removed) when opening the classic checkout page for the first time with seven different UPE payment methods turned on, we will send seven requests to create payment intents since there is nothing in the session storage yet. During the create_payment_intent_ajax()  execution, the storage will be supplemented with intents for each payment method, respectively, allowing for the session storage to be successfully used when users refresh the page or return to the checkout page after traveling to other corners of our product. 

But, the lack of UPE elements loading while switching between radio buttons of different payment methods is explained by the fact that the component is already mounted to the DOM.

### Blocks checkout
Based on the previous research of the classic checkout implementation, we can now bring more light on how switching between different payment methods on the Blocks checkout page works. By leveraging the session storage, we will cut down the amount of new unnecessary payment intents pretty decently since, on each page refresh, no further payment intent will be created, since the existing intent will be taken from the storage. However, the use case of switching back and forth between methods while staying on the page cannot be improved with the help of this mechanism. Well, after the refresh, when we will leverage the session storage, users can switch between methods seamlessly, but only after the page refresh. 

Due to the nature of registerPaymentMethod, we treat each Stripe element as a separate component and render them each time we click on them, which is why paymentMethodsConfig used is not in the up-to-date state.

Having regard to the small study from above, I made the following amendment:
* Leverage the session storage to check if there is a payment intent already which can be used. This improves the performance after the page refresh, but only.

The idea of dealing with the problem of switching elements on the page came to my mind just now, which is based on invoking registerPaymentMethod only once as opposed to [doing it in a loop](https://github.com/Automattic/woocommerce-payments/pull/4681/files#diff-f6bacead8106193d743286ebb7f58e7bad5b05526103306750625a01943cebb5R65-R66), providing the enabledPaymentMethodsConfig to it, and setting up all the enabled payment methods within the one component. 

However, my little hunch here is that we can face integration problems with WooCommerce blocks, both in terms of displaying multiple payment methods as one and supporting the solution in the future without a big headache. In terms of the implementation efforts, I wouldn’t say it is a thing for 8 story points, but the whole upe-fields.js file needs to be reviewed, and some adjustments need to be applied. Let me know what you think about all this.

Reply:
> The mountUPEElement function, among others, [checks if there is no payment intent data related to the particular method in the session storage](https://github.com/Automattic/woocommerce-payments/blob/fix/4603-block-split-upe/client/checkout/classic/upe.js?rgh-link-date=2022-11-02T13%3A42%3A34Z#L231-L233), as you have pointed out before. If there is nothing in the session storage, we contribute to creating a new payment intent with the help of the createIntent Stripe API method.

Just FYI, the lines that you linked to [here](https://github.com/Automattic/woocommerce-payments/blob/fix/4603-block-split-upe/client/checkout/classic/upe.js?rgh-link-date=2022-11-02T13%3A42%3A34Z#L231-L233) aren't related to the intent session storage, but actually the [Pay for Order checkout page](https://woocommerce.com/document/managing-orders/#paying-for-an-order). When a customer pays for an order via this page, there is no cart or checkout session, but the order ID is present as a GET parameter, so we have to use this to look up the order and collect all billable fields from there. 

[This bit](https://github.com/Automattic/woocommerce-payments/blob/fix/4603-block-split-upe/client/checkout/classic/upe.js#L235-L240) is where we check to see if there is any session data present and only create a new payment intent [if there is no intent data found](https://github.com/Automattic/woocommerce-payments/blob/fix/4603-block-split-upe/client/checkout/classic/upe.js#L245)...but I'm pretty sure you knew this already, because it looks like that's basically how you implemented this--so I'll give you the benefit of the doubt and presume that you linked the order ID bit accidentally. Though, this likely redundant explanation is here just in case it was no accident. 😅 

> But, the lack of UPE elements loading while switching between radio buttons of different payment methods is explained by the fact that the component is already mounted to the DOM.

👍 

> However, the use case of switching back and forth between methods while staying on the page cannot be improved with the help of this mechanism.

If there were some way to persist the payment intent after it is created either on the state or by making a request to the backend to store it in the transient cache on the backend, that'd be grand. I don't mind (well, I don't mind so much) that the UPE is remounted every time we switch between payment methods at checkout, but in my testing, it appears that currently a new payment intent is created every time this switch is made! 

I did notice that this is no longer the case, after a page refresh, as you described. However, it would still be really cool if we could figure out a way to update the transient after a payment intent is created (this could be a quick async AJAX call, for example).

> However, my little hunch here is that we can face integration problems with WooCommerce blocks, both in terms of displaying multiple payment methods as one and supporting the solution in the future without a big headache.

🎉 

> In terms of the implementation efforts, I wouldn’t say it is a thing for 8 story points, but the whole upe-fields.js file needs to be reviewed, and some adjustments need to be applied.

Hah, these points are a little arbitrary and that 8 point estimate includes all the work that has already been expended into this issue. That being said, I had a quick scan through your updates and a very quick test running through the checkout flow...and everything looks pretty good to me right now!

I'd rate the basic split-UPE blocks flow as functioning pretty successfully here and would be happy solving how we can be more economical with our intents in a separate issue. Let me know how that sounds or if you have another plan in mind! I think I'd be ready to approve this PR, if you're happy with it, at the earliest convenience (which will probably be Monday, because I'm AFK tomorrow (soz 😞 )). LMK how U feel pls. Cheers!

Reply:
@FangedParakeet a simple one-liner (https://github.com/Automattic/woocommerce-payments/pull/4681/commits/f208c9edd9a9c081e1e682a2ae1d00d3251e8c55) did the trick, and we now create a payment intent only once. Please let me know if it works for you! 

It seems that under the influence of the many elements and mechanisms that make up the split UPE, I was so fascinated that I forgot about this simple solution.